### ISSUE TITLE
Is surface illumination possible?

### ISSUE THREAD

**Issue by Xentha** [nicoco007/BeatSaberCustomAvatars#147]
I've been trying to find a way to cast light from a source, so that I can make it look like my sabers are lighting up my avatar as I swing my arms around.

So far as I know, this is normally done with shaders, but I've yet to find a shader that successfully does it with this mod. Is there something I'm missing, or is this just not currently possible with this mod?

**Comment by nicoco007** [comment]
Shaders are only part of the equation. Basically nothing in Beat Saber actually emits light in a way that shaders can use for illumination; there's some custom directional light stuff going on for environment lighting but that's about it. I've been working on saber & environment lighting for avatars for some time now &ndash; it used to be part of this mod but I extracted it into its own thing since it was causing a lot of noise while I was experimenting. I don't think anyone else has created such a lighting mod but I haven't really done much research.

TL;DR it's not currently possible but I'm experimenting with it.

**Comment by Xentha** [comment]
Oh! You exist again! Went dark for a bit there, glad to see you're still alive.
But, aye, I didn't specifically mean the actual sabers or environment doing light effects, just the ability to attach a light source to the avatar, as if part of the model, that can be placed over the hands to mimic the sabers casting light.

**Comment by Xentha** [comment]
Alright, thank you. I won't have the space to do so until mid september thanks to an upcoming event, but I'll look into it when I can, and post here if I get it working.
